### PR TITLE
documentation: (toy) add lower printf to toy accelerator pass

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: docs/Toy/toy/rewrites/lower_printf_riscv.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: printf.PrintFormatOp, rewriter: PatternRewriter):
+        assert op.format_str.data == "{}"
+        param = op.format_vals[0]
+        memref_typ = param.type
+        assert isa(memref_typ, memref.MemRefType[Attribute])
+        shape = memref_typ.get_shape()
+        if len(shape) == 1:
```
Comment (adutilleul) [inline docs/Toy/toy/rewrites/lower_printf_riscv.py:29]: `nit:` use a match here

File: docs/Toy/toy/rewrites/lower_printf_riscv.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    input := UnrealizedConversionCastOp.get(
+                        (param,), (riscv.IntRegisterType.unallocated(),)
+                    ),
+                    riscv.CustomAssemblyInstructionOp(
+                        "tensor.print2d", (input.results[0], rows.rd, cols.rd), ()
+                    ),
+                ]
+            )
+        else:
+            assert NotImplementedError(f"Cannot print memref with length {len(shape)}")
```
Comment (adutilleul) [inline docs/Toy/toy/rewrites/lower_printf_riscv.py:55]: `nit:` use `raise` here?

**Review by adutilleul — APPROVED** [review]

File: docs/Toy/toy/rewrites/lower_printf_riscv.py (hunk 2 above)
Comment (superlopuh) [reply docs/Toy/toy/rewrites/lower_printf_riscv.py:55]: !!

File: docs/Toy/toy/tests/test_lower_printf.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        with ImplicitBuilder(func.FuncOp("impl", ((), ())).body):
+            m1 = memref.Alloc.get(f64, shape=[2]).memref
+            printf.PrintFormatOp("{}", m1)
+            m2 = memref.Alloc.get(f64, shape=[3, 4]).memref
+            printf.PrintFormatOp("{}", m2)
+
+    return ir_module
+
+
+int_register = riscv.IntRegisterType.unallocated()
```
Comment (adutilleul) [inline docs/Toy/toy/tests/test_lower_printf.py:22]: `INT_REGISTER` ?

**Review by adutilleul — APPROVED** [review]

**Review by kingiler — APPROVED** [review]

**Pull request merged**